Framework/Algorithms/src/FitPeaks.cpp

```diff
@@ -0,0 +1,2301 @@
+ * @param dataws
+ * @param wsindex
+ * @param xmin
+ * @param xmax
+ * @param expected_peak_center
+ * @param observe_peak_width
+ * @param estimate_background
+ * @return
+ */
+double FitPeaks::fitFunctionSD(IAlgorithm_sptr fit,
+                               API::IPeakFunction_sptr peak_function,
+                               API::IBackgroundFunction_sptr bkgd_function,
+                               API::MatrixWorkspace_sptr dataws, size_t wsindex,
+                               double xmin, double xmax,
+                               const double &expected_peak_center,
+                               bool observe_peak_width,
+                               bool estimate_background) {
+
+  // generate peak window
+  std::pair<double, double> peak_window = std::make_pair(xmin, xmax);
+
+  // Estimate background
+  if (estimate_background)
+    estimateBackground(dataws, wsindex, peak_window, bkgd_function);
+  else
+    for (size_t n = 0; n < bkgd_function->nParams(); ++n)
+      bkgd_function->setParameter(n, 0);
+
+  // Estimate peak profile parameter
+  int result =
+      estimatePeakParameters(dataws, wsindex, peak_window, peak_function,
+                             bkgd_function, observe_peak_width);
+  if (result != GOOD)
+    peak_function->setCentre(expected_peak_center);
+
+  // Create the composition function
+  CompositeFunction_sptr comp_func =
+      boost::make_shared<API::CompositeFunction>();
+  comp_func->addFunction(peak_function);
+  comp_func->addFunction(bkgd_function);
+  IFunction_sptr fitfunc = boost::dynamic_pointer_cast<IFunction>(comp_func);
+
+  // Set the properties
+  fit->setProperty("Function", fitfunc);
+  fit->setProperty("InputWorkspace", dataws);
+  fit->setProperty("WorkspaceIndex", static_cast<int>(wsindex));
+  fit->setProperty("MaxIterations", 50); // magic number
+  fit->setProperty("StartX", xmin);
+  fit->setProperty("EndX", xmax);
+
+  if (constrain_peaks_position_) {
+    // set up a constraint on peak position
+    double peak_center = peak_function->centre();
+    double peak_width = peak_function->fwhm();
+    std::stringstream peak_center_constraint;
+    peak_center_constraint << (peak_center - 0.5 * peak_width) << " < f0."
+                           << peak_function->getCentreParameterName() << " < "
+                           << (peak_center + 0.5 * peak_width);
+
+    // set up a constraint on peak height
+
+    fit->setProperty("Constraints", peak_center_constraint.str());
+  }
+
+  // Execute fit and get result of fitting background
+  // m_sstream << "FitSingleDomain: " << fit->asString() << ".\n";
+  g_log.debug() << "[E1201] FitSingleDomain Before fitting, Fit function: "
+                << fit->asString() << "\n";
+
+  fit->executeAsChildAlg();
+
+  g_log.debug() << "[E1202] FitSingleDomain After fitting, Fit function: "
+                << fit->asString() << "\n";
+
+  if (!fit->isExecuted()) {
+    g_log.error("Fitting peak SD (single domain) failed to execute.");
```

Comment:
Logging an error then throwing the exception is unnecessary. The caller will catch the exception and log it to the error channel then. Remove the `g_log.error()`.